mlkem/reduce.c

```diff
@@ -41,26 +41,41 @@ int16_t montgomery_reduce(int32_t a) {
   // Replace C -> C * q in the above and estimate
   // q / 2^17 < 0.0254.
 
-  uint16_t u;
-  int16_t t;
-  // Compute a*q^{-1} mod 2^16 in unsigned representatives
-  u = (uint16_t)a * QINV;
+  // QINV == -3327 converted to uint16_t == -3327 + 65536 == 62209
+  const uint32_t UQINV = 62209;
+
+  //  Compute a*q^{-1} mod 2^16 in unsigned representatives
+  const uint16_t a_inverted =
+      (uint16_t)((((uint32_t)(a & 0xFFFF)) * UQINV) & 0xFFFF);
```

Comment:
Are all those annotations really necessary? It seems like CBMC would complain about _any_ situation where you multiply two `uint16_t` into a `uint16_t` because the multiplication happens promoted, and then gets truncated?

Reply:
They "& 0xFFFF" masks are required to convince CBMC that the "*" operator doesn't overflow. Removing either causes multiple proof failures.

Reply:
So _any_ uint16 x uint16 -> uint16 multiplication in needs this contortion to go through CBMC?